mpf/devices/blinkenlight.py

```diff
@@ -0,0 +1,80 @@
+"""Contains the Blinkenlight class."""
+import asyncio
+from functools import partial
+
+from typing import Set, Dict, List, Tuple, Any
+
+from mpf.core.rgb_color import RGBColor, ColorException
+from mpf.core.system_wide_device import SystemWideDevice
+from mpf.core.delays import DelayManager
+from mpf.core.device_monitor import DeviceMonitor
+
+@DeviceMonitor("num_colors", "current_color", "light")
+class Blinkenlight(SystemWideDevice):
+
+    """A light that alternates between several different colors."""
+
+    config_section = 'blinkenlights'
+    collection = 'blinkenlights'
+    class_label = 'blinkenlight'
+
+    def __init__(self, machine, name):
+        """Initialise blinkenlight."""
+        super().__init__(machine, name)
+
+        # self.each_duration = 0
+        # self.off_duration = 0
+        # self.off_when_multiple = False
+        # self.light = None   # type: Light
+        self._colors = []
+        self.delay = DelayManager(machine)
+        self._color_i = 0
+        self._num_colors = 0
+        # Add this device to the blinkenlight section
+        self.machine.blinkenlights[name] = self
+
+    @property
+    def colors(self):
+        return self._colors
+
+    @property
+    def num_colors(self):
+        # self.info_log('num_colors queried - returning {}'.format(len(self._colors)))
+        return len(self._colors)
+
+    @num_colors.setter
+    def num_colors(self, newValue):
+        # self.info_log('num_colors set')
+        self._num_colors = newValue
+
+    @property
+    def current_color(self):
+        if self._color_i >= len(self._colors):
+            return None
+        if self._color_i < 0:
+            return None
+        return self._colors[self._color_i][0]
+
+    @property
+    def light(self):
+        return self.config['light']
+
+    def _restart(self):
+        self._color_i = 0
+        self.delay.clear()
+        self.config['light'].color(RGBColor("off"))
+        self._perform_step()
+
+    def _perform_step(self):
+        if len(self._colors) == 0:
```

Comment:
I would turn this into a coroutine. We just need to figure out how we synchonize all blinkenlights. Maybe always start at sync_ms xx. Would that work? That would slightly change the logic. I looked at some sterns and it seems that they got up to 2 slots (i.e. on star trek). If both are filled the light will toggle between them. If one is filled it will turn off. Lights seem to be mostly in sync.

Reply:
I'll have to think about this.  Are you suggesting that each color change is in sync (so that each blinkenlight will change from one color to the next in sync with the others), or that each blinkenlight starts/finishes its entire cycle in sync will all other blinkenlights?  In the second case, I don't know how it would work if one blinkenlight had 2 colors and another has 6, since it would take the 6 color blinkenlight three times as long to cycle through its colors as the 2 color blinkenlight, so they couldn't stay in sync.

Reply:
I would imagine to sync the cycles. One idea would be to run them at a fixed duration and fill up the remaining slots with black (or transparency). This seems to be the case for sterns.

I see that this would be an issue with variable length cycles. Alternatively, we could only sync same length cycles (by using sync_ms = cycle length). That might still work well for the common case: green + off and green + red would sync but green + red + yellow would not sync. Guess that would be ok.

What do you think?

Reply:
I’m thinking I could make both options happen. Either you could specify a cycle_duration=1 for example, so that each cycle is 1 second regardless of how many colors there are, or each_duration=1 would specify that each color lasts 1 second. And then the sync_ms would sync accordingly (either on a cycle or color basis). Would that work?

Reply:
So with Stern, Game of Thrones has 3 colors that can be on a single light at a time.  I have a slow mo video, if you want.  I could upload and share the link.

The way it works is all are the same duration.  Lets say 1 for example.  Then each light has its cycle of what it can do.  LED1 has 1 color (R), LED2 has 2 colors (R,B), and LED3 has 3 colors (R,B,G).  They all loop through their colors and go to off at the end before restarting.  So here is a time breakdown.

Time 0: LED1 = R, LED2 = R, LED3 = R
Time 1: LED1 = Off, LED2 = B, LED3 = B
Time 2: LED1 = R, LED2 = Off, LED3 = G
Time 3: LED1= Off, LED2 = R, LED3 = Off

This would repeat.  So they aren't all showing red at once, or anything like that.  But that all shift at the exact same time.  So we would need a sync_ms for all blinkenlights (not each individual light), and a fixed duration for all lights (not each individual light).  This would mimic Stern to make it seamless to the user, as you have things all shifting simultaneously, instead of randomly.

Reply:
Thanks for checking that. We should support it this way.

Can you check one more thing for me: If you got two lights with R+B+off are those always in sync? I mean does both show red at the same time? I would guess so.

Basically, the above is what we should get with sync ms based on cycle length. Maybe, we can hack this together in the next few days.

Reply:
Yes, all binkenlights are exactly on the same sync.  So they all hit color 1 at the exact same time, color 2, etc.  So if two lights have the same values (R, B, Off), they would be perfectly in sync.  Then when you hit the R on shot 1, that gets removed from the loop for that light.  So then they aren't showing the same colors at the same time (1 showing B, Off and the other showing R, B, Off), but the sync_ms for transition is identical.  Here is a link to a video showing this on Game of Thrones: https://pindev.slack.com/files/U01GQ857T4L/F02A78ME7KR/img_8126.mov

Reply:
Thanks. I believe the way I've outlined would work then. What you're describing for how GoT works would be just like setting color_duration, so that each color on a particular light lasts for a given amount of time. The sync issue still might need some thought, since a sync_ms on each blinkenlight wouldn't ensure that they all started their cycle at the same time (for blinkenlights with the same number of colors). For example, if I have 2 blinkenlights and they're each set with color_duration=500ms, and sync_ms=1s, this would ensure they started at the same time if they each had 2 colors. But if they each had 4 colors, they could be a half-cycle off from each other.

Reply:
That's a good point.  So might need to be at a higher level.  let me do some "research" on Game of Thrones.  I can get a mode started, and then add shots by progressing the modes.  I will see if it then syncs up with all shots or not.  I have a suspicious the answer will be yes.

But regardless of how Stern implemented, it would be nice to have them all start from the same place.  But with long duration, that becomes a bit of lag between when the event occurs and the lights add that color.

Reply:
How about this:

Blibkenlight_player figures out the sync_ms automatically, based on the current cycle length of the blinkenlight (I.e., the number of colors the light currently has if color_duration is set, or else the cycle_duration value if it is set). This is probably what Jan was suggesting earlier. Based on this, the blinkenlight_player could figure out which color should be on at any given time so that it wouldn't have to wait till the beginning of its cycle if the sync_ms happens to be very large when it starts.

In this case, there would be no user-configurable sync_ms option for blinkenlights, since it would be calculated and synced automatically.

However this does bring up another issue. This would require that all colors are added to the blinkenlight in the same order, or else some other ordering would be required. If one blinkenlight had its colors added R then G but reversed for the other blinkenlight, then these would flash alternating colors.

Reply:
In the latest commit, I addressed all these.  I added different parameters:

blinkenlights:
  my_blinkenlight:
    color_duration: 1s
    off_when_multiple: false
    light: l_stark_sigil
    priority: 1000
  my_blinkenlight2:
    cycle_duration: 1s
    off_when_multiple: false
    light: l_martell_sigil
    priority: 1000

Now you can either specify color_duration or cycle_duration (you have to specify one or the other, but not both - this is validated in code).  Sync across multiple blinkenlights is achieved based on the logic I outlined above (i.e., the cycle duration is automatically calculated and a "virtual" sync_ms is calculated based on the length of the cycle).  When the blinkenlight starts, it doesn't have to wait till the beginning of the next cycle (as it would with a typical sync_ms), but will instead figure out which color should be on at that point and go from there.

Also I added a priority.  This defaults to 0 if not specified.

Reply:
That sounds awesome!

> Sync across multiple blinkenlights is achieved based on the logic I outlined above (i.e., the cycle duration is automatically calculated and a "virtual" sync_ms is calculated based on the length of the cycle). When the blinkenlight starts, it doesn't have to wait till the beginning of the

We have to check how that looks. Might give us weird effects im some cases. One slightly more complex alternative would be to always finish the current cycle and only considering changes when the loop starts. Guess that would be something we might add later.

Reply:
Yes this could be weird in some cases but it works in a predictable manner. For example, if a blibkenlight has 2 colors (let's say red and green), so it blinks red, green, red, green. If green is removed by a mode while the green light is lit, then for the rest of green's "slot" time, it would the light would be off since now the blinkenlight should flash red, off, red, off. This could cause it to look like it flickers but it's intended behavior in this case.